File: examples/host-demo-python/deluge.py
"""
Deluge USB Serial Protocol — Python library

Message framing:
    [length_lo : 1][length_hi : 1][type : 1][data : N]
    length = 1 (type) + N (data), NOT including the 2 length bytes.

See src/controller/usb_serial.h for the canonical message type definitions.
"""

from __future__ import annotations

import struct
import time
from dataclasses import dataclass
from enum import IntEnum
from typing import Optional

import serial
import serial.tools.list_ports

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

DELUGE_VID = 0x16D0
DELUGE_PID = 0x0EDA

# OLED display dimensions
OLED_WIDTH: int = 128
OLED_HEIGHT: int = 48
OLED_PAGES: int = OLED_HEIGHT // 8              # 6
OLED_FRAMEBUFFER_SIZE: int = OLED_PAGES * OLED_WIDTH  # 768
# Top rows hidden behind the display bezel (matches OLED_MAIN_TOPMOST_PIXEL)
OLED_TOPMOST_PIXEL: int = 5
OLED_VISIBLE_HEIGHT: int = OLED_HEIGHT - OLED_TOPMOST_PIXEL  # 43

# Pad grid dimensions
PAD_COLS: int = 18
PAD_ROWS: int = 8

# ---------------------------------------------------------------------------
# Message types
# ---------------------------------------------------------------------------

class MsgFrom(IntEnum):
    """Messages FROM Deluge TO Host."""
    PAD_PRESSED      = 0x01
    PAD_RELEASED     = 0x02
    BUTTON_PRESSED   = 0x03
    BUTTON_RELEASED  = 0x04
    ENCODER_ROTATED  = 0x05
    ENCODER_PRESSED  = 0x06
    ENCODER_RELEASED = 0x07
    VERSION          = 0x10
    PONG             = 0x11
    READY            = 0x12
    ERROR            = 0x13


class MsgTo(IntEnum):
    """Messages TO Deluge FROM Host."""
    UPDATE_DISPLAY    = 0x20
    CLEAR_DISPLAY     = 0x21
    SET_PAD_RGB       = 0x22
    CLEAR_ALL_PADS    = 0x23
    SET_LED           = 0x24
    SET_CV            = 0x25
    SET_GATE          = 0x26
    SET_ALL_PADS      = 0x27
    SET_KNOB_INDICATOR = 0x28
    SET_SYNCED_LED    = 0x29
    GET_VERSION       = 0x30
    PING              = 0x31

# ---------------------------------------------------------------------------
# Event types
# ---------------------------------------------------------------------------

@dataclass
class PadPressed:
    col: int
    row: int

@dataclass
class PadReleased:
    col: int
    row: int

@dataclass
class ButtonPressed:
    id: int

@dataclass
class ButtonReleased:
    id: int

@dataclass
class EncoderRotated:
    id: int
    delta: int  # signed; positive = clockwise

@dataclass
class EncoderPressed:
    id: int

@dataclass
class EncoderReleased:
    id: int

@dataclass
class VersionEvent:
    major: int
    minor: int
    patch: int

@dataclass
class Pong:
    pass

@dataclass
class Ready:
    pass

@dataclass
class ErrorEvent:
    message: str

# Union type alias for type hints
DelugeEvent = (
    PadPressed | PadReleased |
    ButtonPressed | ButtonReleased |
    EncoderRotated | EncoderPressed | EncoderReleased |
    VersionEvent | Pong | Ready | ErrorEvent
)

# ---------------------------------------------------------------------------
# Deluge connection
# ---------------------------------------------------------------------------

class DelugeError(Exception):
    """Raised for protocol-level errors."""


class Deluge:
    """
    Connection to a Deluge over USB CDC serial.

    Usage::

        with Deluge.open() as d:
            d.ping()
            d.set_pad(0, 0, 255, 0, 0)
            event = d.try_read_event()
    """

    def __init__(self, port: serial.Serial) -> None:
        self._port = port

    # ------------------------------------------------------------------
    # Construction
    # ------------------------------------------------------------------

    @classmethod
    def open(cls) -> "Deluge":
        """
        Auto-detect and connect to the Deluge (VID:PID 16D0:0EDA).
        Raises RuntimeError if not found.
        """
        ports = list(serial.tools.list_ports.comports())

        for info in ports:
            if info.vid == DELUGE_VID and info.pid == DELUGE_PID:
                print(f"Found Deluge on {info.device}")
                return cls._open_device(info.device)

        if not ports:
            raise RuntimeError(
                "No serial ports found. Is the Deluge connected via USB?"
            )

        print("Available serial ports:")
        for p in ports:
            print(f"  {p.device}  vid={p.vid:#06x} pid={p.pid:#06x}  desc={p.description}")
        raise RuntimeError(
            "Deluge not found (VID:PID 16D0:0EDA). "
            "Is the USB controller firmware flashed?"
        )

    @classmethod
    def open_port(cls, name: str) -> "Deluge":
        """Connect to a specific serial port by name."""
        return cls._open_device(name)

    @classmethod
    def _open_device(cls, name: str) -> "Deluge":
        port = serial.Serial(
            name,
            baudrate=115_200,
            timeout=0.5,       # 500 ms read timeout (matches Rust)
        )
        # Assert DTR so TinyUSB's tud_cdc_connected() returns True.
        try:
            port.dtr = True
        except Exception as exc:
            print(f"Warning: could not set DTR: {exc}")
        # Small delay for the device to process the line state change.
        time.sleep(0.1)
        return cls(port)

    # ------------------------------------------------------------------
    # Context manager support
    # ------------------------------------------------------------------

    def __enter__(self) -> "Deluge":
        return self

    def __exit__(self, *_) -> None:
        self.close()

    def close(self) -> None:
        """Close the serial port."""
        if self._port and self._port.is_open:
            self._port.close()

    # ------------------------------------------------------------------
    # Low-level send / receive
    # ------------------------------------------------------------------

    def _send_raw(self, msg_type: int, data: bytes | bytearray = b"") -> None:
        """
        Frame and transmit one message.

        Wire layout: [length_lo][length_hi][type][data...]
        where length = 1 + len(data).
        """
        length = 1 + len(data)
        frame = struct.pack("<H", length) + bytes([msg_type]) + bytes(data)

        written = 0
        deadline = time.monotonic() + 2.0
        while written < len(frame):
            try:
                n = self._port.write(frame[written:])
                if n:
                    written += n
            except serial.SerialTimeoutException:
                if time.monotonic() >= deadline:
                    raise DelugeError("Write timed out sending to Deluge")
                time.sleep(0.005)

    def _read_exact(self, n: int, allow_timeout_on_first: bool = False) -> Optional[bytes]:
        """
        Read exactly *n* bytes.
        Returns None if no data arrives at all (clean timeout) and
        allow_timeout_on_first is True; otherwise raises on timeout.
        """
        buf = bytearray()
        first = True
        while len(buf) < n:
            chunk = self._port.read(n - len(buf))
            if chunk:
                buf.extend(chunk)
                first = False
            else:
                # No data returned within the port timeout
                if first and allow_timeout_on_first:
                    return None
                if not first:
                    # Partial read — keep retrying
                    continue
                return None
        return bytes(buf)

    def try_read_event(self) -> Optional[DelugeEvent]:
        """
        Non-blocking read of one event.  Returns None if no data is available.
        """
        # Fast check — avoid waiting for the full port timeout when idle.
        if self._port.in_waiting < 3:
            return None

        # Read 2-byte length header
        hdr = self._read_exact(2, allow_timeout_on_first=True)
        if hdr is None:
            return None
        (msg_len,) = struct.unpack("<H", hdr)

        if msg_len == 0 or msg_len > 1024:
            raise DelugeError(f"Invalid message length: {msg_len}")

        payload = self._read_exact(msg_len)
        if payload is None:
            raise DelugeError("Timeout reading message payload after header")

        msg_type = payload[0]
        data = payload[1:]

        try:
            mt = MsgFrom(msg_type)
        except ValueError:
            raise DelugeError(f"Unknown message type: {msg_type:#04x}")

        if mt == MsgFrom.PAD_PRESSED and len(data) >= 2:
            return PadPressed(col=data[0], row=data[1])
        elif mt == MsgFrom.PAD_RELEASED and len(data) >= 2:
            return PadReleased(col=data[0], row=data[1])
        elif mt == MsgFrom.BUTTON_PRESSED and len(data) >= 1:
            return ButtonPressed(id=data[0])
        elif mt == MsgFrom.BUTTON_RELEASED and len(data) >= 1:
            return ButtonReleased(id=data[0])
        elif mt == MsgFrom.ENCODER_ROTATED and len(data) >= 2:
            delta = struct.unpack("b", bytes([data[1]]))[0]  # signed byte
            return EncoderRotated(id=data[0], delta=delta)
        elif mt == MsgFrom.ENCODER_PRESSED and len(data) >= 1:
            return EncoderPressed(id=data[0])
        elif mt == MsgFrom.ENCODER_RELEASED and len(data) >= 1:
            return EncoderReleased(id=data[0])
        elif mt == MsgFrom.VERSION and len(data) >= 3:
            return VersionEvent(major=data[0], minor=data[1], patch=data[2])
        elif mt == MsgFrom.PONG:
            return Pong()
        elif mt == MsgFrom.READY:
            return Ready()
        elif mt == MsgFrom.ERROR:
            return ErrorEvent(message=data.decode("utf-8", errors="replace"))
        else:
            raise DelugeError(
                f"Message {msg_type:#04x} with unexpected data length {len(data)}"
            )

    # ------------------------------------------------------------------
    # High-level commands ← matches Rust Deluge impl methods
    # ------------------------------------------------------------------

    def ping(self, timeout: float = 1.0) -> None:
        """Send a ping and wait for a pong response."""
        self._send_raw(MsgTo.PING)
        deadline = time.monotonic() + timeout
        while True:
            if time.monotonic() >= deadline:
                raise DelugeError("Ping timed out — no pong received")
            event = self.try_read_event()
            if isinstance(event, Pong):
                return

    def get_version(self, timeout: float = 1.0) -> tuple[int, int, int]:
        """Request firmware version; returns (major, minor, patch)."""
        self._send_raw(MsgTo.GET_VERSION)
        deadline = time.monotonic() + timeout
        while True:
            if time.monotonic() >= deadline:
                raise DelugeError("get_version timed out")
            event = self.try_read_event()
            if isinstance(event, VersionEvent):
                return (event.major, event.minor, event.patch)

    def set_pad(self, col: int, row: int, r: int, g: int, b: int) -> None:
        """Set a single pad LED colour (col 0-17, row 0-7)."""
        self._send_raw(MsgTo.SET_PAD_RGB, bytes([col, row, r, g, b]))

    def set_all_pads(self, colors: list[tuple[int, int, int]]) -> None:
        """
        Bulk RGB pad update.

        *colors* must be a flat col-major list of (r, g, b) tuples with
        PAD_COLS * PAD_ROWS = 144 entries:
            colors[col * PAD_ROWS + row] = (r, g, b)
        """
        if len(colors) != PAD_COLS * PAD_ROWS:
            raise ValueError(
                f"set_all_pads expects {PAD_COLS * PAD_ROWS} entries, got {len(colors)}"
            )
        buf = bytearray(PAD_COLS * PAD_ROWS * 3)
        for i, (r, g, b) in enumerate(colors):
            buf[i * 3]     = r
            buf[i * 3 + 1] = g
            buf[i * 3 + 2] = b
        self._send_raw(MsgTo.SET_ALL_PADS, buf)

    def clear_pads(self) -> None:
        """Turn off all pad LEDs."""
        self._send_raw(MsgTo.CLEAR_ALL_PADS)

    def update_display(self, framebuffer: bytes | bytearray) -> None:
        """
        Push a full OLED framebuffer.

        *framebuffer* must be exactly OLED_FRAMEBUFFER_SIZE (768) bytes.
        Encoding: page-major (6 pages × 128 columns), LSB of each byte = top
        pixel of that page column.
        """
        if len(framebuffer) != OLED_FRAMEBUFFER_SIZE:
            raise ValueError(
                f"framebuffer must be {OLED_FRAMEBUFFER_SIZE} bytes, "
                f"got {len(framebuffer)}"
            )
        self._send_raw(MsgTo.UPDATE_DISPLAY, framebuffer)

    def clear_display(self) -> None:
        """Clear the OLED display."""
        self._send_raw(MsgTo.CLEAR_DISPLAY)

    def set_led(self, index: int, on: bool) -> None:
        """Set an indicator LED on or off (index 0–35)."""
        self._send_raw(MsgTo.SET_LED, bytes([index, int(on)]))

    def set_knob_indicator(self, which: int, levels: tuple[int, int, int, int]) -> None:
        """
        Set the 4-LED gold-knob indicator bar.

        *which*: 0 = MOD_0 (bottom knob), 1 = MOD_1 (top knob).
        *levels*: brightness for each LED segment (0 = off, 255 = full).
        """
        self._send_raw(MsgTo.SET_KNOB_INDICATOR, bytes([which, *levels]))

    def set_synced_led(self, on: bool) -> None:
        """Control the external-sync LED on the rear panel."""
        self._send_raw(MsgTo.SET_SYNCED_LED, bytes([int(on)]))

    def set_cv(self, channel: int, value: int) -> None:
        """
        Set a CV output voltage.

        *channel*: 0–3.
        *value*: 0–65535, linear across 0–10 V.
        """
        self._send_raw(MsgTo.SET_CV, bytes([channel, (value >> 8) & 0xFF, value & 0xFF]))

    def set_gate(self, channel: int, on: bool) -> None:
        """Set a gate output (+5 V when on). channel: 0–3."""
        self._send_raw(MsgTo.SET_GATE, bytes([channel, int(on)]))

    # ------------------------------------------------------------------
    # Utilities
    # ------------------------------------------------------------------

    def drain_events(self) -> list[DelugeEvent]:
        """Read and return all currently buffered events."""
        events: list[DelugeEvent] = []
        while True:
            event = self.try_read_event()
            if event is None:
                break
            events.append(event)
        return events


# ---------------------------------------------------------------------------
# OLED framebuffer helpers
# ---------------------------------------------------------------------------

def oled_set_pixel(fb: bytearray, x: int, y: int) -> None:
    """Set pixel (x, y) in a page-encoded OLED framebuffer."""
    if x < 0 or x >= OLED_WIDTH or y < 0 or y >= OLED_HEIGHT:
        return
    page = y // 8
    bit  = y % 8
    fb[page * OLED_WIDTH + x] |= 1 << bit


def oled_clear_pixel(fb: bytearray, x: int, y: int) -> None:
    """Clear pixel (x, y) in a page-encoded OLED framebuffer."""
    if x < 0 or x >= OLED_WIDTH or y < 0 or y >= OLED_HEIGHT:
        return
    page = y // 8
    bit  = y % 8
    fb[page * OLED_WIDTH + x] &= ~(1 << bit)


def oled_new_framebuffer() -> bytearray:
    """Return a zeroed OLED framebuffer."""
    return bytearray(OLED_FRAMEBUFFER_SIZE)


def oled_draw_filled_circle(fb: bytearray, cx: int, cy: int, r: int) -> None:
    """Draw a filled circle at (cx, cy) with radius r."""
    for dy in range(-r, r + 1):
        for dx in range(-r, r + 1):
            if dx * dx + dy * dy <= r * r:
                oled_set_pixel(fb, cx + dx, cy + dy)


def oled_draw_string(fb: bytearray, x0: int, y0: int, text: str) -> None:
    """Draw a string using the built-in 5×7 bitmap font."""
    x = x0
    for ch in text:
        _oled_draw_char(fb, x, y0, ch)
        x += 6  # 5 pixels + 1 gap


def _oled_draw_char(fb: bytearray, x0: int, y0: int, ch: str) -> None:
    glyph = _FONT_5X7.get(ch.upper())
    if glyph is None:
        return
    for dx, col_bits in enumerate(glyph):
        for bit in range(7):
            if col_bits & (1 << bit):
                oled_set_pixel(fb, x0 + dx, y0 + bit)


# 5×7 bitmap font — each char is 5 column bytes (bit 0 = top row)
_FONT_5X7: dict[str, tuple[int, int, int, int, int]] = {
    '0': (0x3E, 0x51, 0x49, 0x45, 0x3E),
    '1': (0x00, 0x42, 0x7F, 0x40, 0x00),
    '2': (0x42, 0x61, 0x51, 0x49, 0x46),
    '3': (0x21, 0x41, 0x45, 0x4B, 0x31),
    '4': (0x18, 0x14, 0x12, 0x7F, 0x10),
    '5': (0x27, 0x45, 0x45, 0x45, 0x39),
    '6': (0x3C, 0x4A, 0x49, 0x49, 0x30),
    '7': (0x01, 0x71, 0x09, 0x05, 0x03),
    '8': (0x36, 0x49, 0x49, 0x49, 0x36),
    '9': (0x06, 0x49, 0x49, 0x29, 0x1E),
    ':': (0x00, 0x36, 0x36, 0x00, 0x00),
    'F': (0x7F, 0x09, 0x09, 0x09, 0x01),
    'P': (0x7F, 0x09, 0x09, 0x09, 0x06),
    'R': (0x7F, 0x09, 0x19, 0x29, 0x46),
    'E': (0x7F, 0x49, 0x49, 0x49, 0x41),
    'S': (0x46, 0x49, 0x49, 0x49, 0x31),
    'D': (0x7F, 0x41, 0x41, 0x22, 0x1C),
    'B': (0x7F, 0x49, 0x49, 0x49, 0x36),
    'A': (0x7E, 0x11, 0x11, 0x11, 0x7E),
    'L': (0x7F, 0x40, 0x40, 0x40, 0x40),
    'O': (0x3E, 0x41, 0x41, 0x41, 0x3E),
    'N': (0x7F, 0x04, 0x08, 0x10, 0x7F),
    'T': (0x01, 0x01, 0x7F, 0x01, 0x01),
    ' ': (0x00, 0x00, 0x00, 0x00, 0x00),
    'I': (0x00, 0x41, 0x7F, 0x41, 0x00),
    'C': (0x3E, 0x41, 0x41, 0x41, 0x22),
    'G': (0x3E, 0x41, 0x49, 0x49, 0x3A),
    'H': (0x7F, 0x08, 0x08, 0x08, 0x7F),
    'K': (0x7F, 0x08, 0x14, 0x22, 0x41),
    'M': (0x7F, 0x02, 0x0C, 0x02, 0x7F),
    'U': (0x3F, 0x40, 0x40, 0x40, 0x3F),
    'V': (0x1F, 0x20, 0x40, 0x20, 0x1F),
    'W': (0x3F, 0x40, 0x38, 0x40, 0x3F),
    'X': (0x63, 0x14, 0x08, 0x14, 0x63),
    'Y': (0x07, 0x08, 0x70, 0x08, 0x07),
    'Z': (0x61, 0x51, 0x49, 0x45, 0x43),
}

File: examples/host-demo/src/deluge.rs
// Deluge USB Serial Protocol Implementation
//
// Message format: [length_lo:1][length_hi:1][type:1][data:N]
// Length = 1 (type byte) + N (data bytes), does NOT include the 2 length bytes.

use anyhow::{bail, Context, Result};
use std::io::{Read, Write};
use std::time::{Duration, Instant};

// ── Message types FROM Deluge TO Host ────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum MsgFrom {
    PadPressed = 0x01,
    PadReleased = 0x02,
    ButtonPressed = 0x03,
    ButtonReleased = 0x04,
    EncoderRotated = 0x05,
    EncoderPressed = 0x06,
    EncoderReleased = 0x07,
    Version = 0x10,
    Pong = 0x11,
    Ready = 0x12,
    Error = 0x13,
}

impl MsgFrom {
    fn from_byte(b: u8) -> Option<Self> {
        match b {
            0x01 => Some(Self::PadPressed),
            0x02 => Some(Self::PadReleased),
            0x03 => Some(Self::ButtonPressed),
            0x04 => Some(Self::ButtonReleased),
            0x05 => Some(Self::EncoderRotated),
            0x06 => Some(Self::EncoderPressed),
            0x07 => Some(Self::EncoderReleased),
            0x10 => Some(Self::Version),
            0x11 => Some(Self::Pong),
            0x12 => Some(Self::Ready),
            0x13 => Some(Self::Error),
            _ => None,
        }
    }
}

// ── Message types TO Deluge FROM Host ────────────────────────────────

#[repr(u8)]
enum MsgTo {
    UpdateDisplay = 0x20,
    ClearDisplay = 0x21,
    SetPadRgb = 0x22,
    ClearAllPads = 0x23,
    #[allow(dead_code)]
    SetLed = 0x24,
    #[allow(dead_code)]
    SetCv = 0x25,
    #[allow(dead_code)]
    SetGate = 0x26,
    ClearAllLeds = 0x2A,
    SetAllPads = 0x27,
    SetKnobIndicator = 0x28,
    SetSyncedLed = 0x29,
    GetVersion = 0x30,
    Ping = 0x31,
}

// ── Parsed events from the Deluge ────────────────────────────────────

#[derive(Debug, Clone)]
pub enum DelugeEvent {
    PadPressed { col: u8, row: u8 },
    PadReleased { col: u8, row: u8 },
    ButtonPressed { id: u8 },
    ButtonReleased { id: u8 },
    EncoderRotated { id: u8, delta: i8 },
    EncoderPressed { id: u8 },
    EncoderReleased { id: u8 },
    Version { major: u8, minor: u8, patch: u8 },
    Pong,
    Ready,
    Error(String),
}

// ── Deluge connection ────────────────────────────────────────────────

pub struct Deluge {
    port: Box<dyn serialport::SerialPort>,
}

/// OLED display dimensions
pub const OLED_WIDTH: usize = 128;
pub const OLED_HEIGHT: usize = 48;
pub const OLED_PAGES: usize = OLED_HEIGHT / 8; // 6
pub const OLED_FRAMEBUFFER_SIZE: usize = OLED_PAGES * OLED_WIDTH; // 768
/// Top rows of the framebuffer that are hidden behind the display bezel.
/// Matches OLED_MAIN_TOPMOST_PIXEL in cpu_specific.h.
pub const OLED_TOPMOST_PIXEL: usize = 5;
/// Visible pixel rows (framebuffer rows OLED_TOPMOST_PIXEL..OLED_HEIGHT).
pub const OLED_VISIBLE_HEIGHT: usize = OLED_HEIGHT - OLED_TOPMOST_PIXEL; // 43

/// Pad grid dimensions
pub const PAD_COLS: u8 = 18;
pub const PAD_ROWS: u8 = 8;

impl Deluge {
    /// Auto-detect and open the Deluge serial port.
    pub fn open() -> Result<Self> {
        let ports = serialport::available_ports().context("Failed to list serial ports")?;

        // Look for Deluge VID:PID = 16D0:0EDA
        for port_info in &ports {
            if let serialport::SerialPortType::UsbPort(usb) = &port_info.port_type {
                if usb.vid == 0x16D0 && usb.pid == 0x0EDA {
                    println!("Found Deluge on {}", port_info.port_name);
                    let mut port = serialport::new(&port_info.port_name, 115200)
                        .timeout(Duration::from_millis(500))
                        .open()
                        .with_context(|| {
                            format!("Failed to open serial port {}", port_info.port_name)
                        })?;
                    // Assert DTR so TinyUSB's tud_cdc_connected() returns true
                    if let Err(e) = port.write_data_terminal_ready(true) {
                        eprintln!("Warning: could not set DTR: {e}");
                    }
                    // Small delay for the device to process the line state change
                    std::thread::sleep(Duration::from_millis(100));
                    return Ok(Self { port });
                }
            }
        }

        // Fallback: list available ports and give a helpful message
        if ports.is_empty() {
            bail!("No serial ports found. Is the Deluge connected via USB?");
        }
        println!("Available serial ports:");
        for p in &ports {
            println!("  {} ({:?})", p.port_name, p.port_type);
        }
        bail!("Deluge not found (VID:PID 16D0:0EDA). Is the USB controller firmware flashed?");
    }

    /// Open a specific serial port by name.
    pub fn open_port(name: &str) -> Result<Self> {
        let mut port = serialport::new(name, 115200)
            .timeout(Duration::from_millis(500))
            .open()
            .with_context(|| format!("Failed to open serial port {}", name))?;
        // Assert DTR so TinyUSB's tud_cdc_connected() returns true
        if let Err(e) = port.write_data_terminal_ready(true) {
            eprintln!("Warning: could not set DTR: {e}");
        }
        std::thread::sleep(Duration::from_millis(100));
        Ok(Self { port })
    }

    // ── Low-level send/receive ───────────────────────────────────────

    fn send_raw(&mut self, msg_type: u8, data: &[u8]) -> Result<()> {
        let length = 1u16 + data.len() as u16; // type + data
        let mut buf = Vec::with_capacity(3 + data.len());
        buf.push((length & 0xFF) as u8);
        buf.push((length >> 8) as u8);
        buf.push(msg_type);
        buf.extend_from_slice(data);
        // write_all can return ERROR_SEM_TIMEOUT (121) on Windows CDC — retry on timeout
        let mut written = 0usize;
        let deadline = Instant::now() + Duration::from_secs(2);
        while written < buf.len() {
            match self.port.write(&buf[written..]) {
                Ok(n) => written += n,
                Err(e) if Self::is_timeout(&e) => {
                    if Instant::now() >= deadline {
                        return Err(e).context("Write timed out sending to Deluge");
                    }
                    std::thread::sleep(Duration::from_millis(5));
                    continue;
                }
                Err(e) => return Err(e.into()),
            }
        }
        Ok(())
    }

    /// Check if an I/O error is a timeout (handles Windows ERROR_SEM_TIMEOUT = 121).
    fn is_timeout(e: &std::io::Error) -> bool {
        e.kind() == std::io::ErrorKind::TimedOut
            || e.raw_os_error() == Some(121) // ERROR_SEM_TIMEOUT on Windows
    }

    /// Read exactly `n` bytes, returning `None` if no data arrives at all (timeout).
    /// Once the first byte is received, keeps reading until all bytes arrive.
    fn read_exact_or_timeout(&mut self, buf: &mut [u8]) -> Result<Option<()>> {
        let mut filled = 0usize;
        while filled < buf.len() {
            match self.port.read(&mut buf[filled..]) {
                Ok(0) => bail!("Serial port closed unexpectedly"),
                Ok(n) => filled += n,
                Err(e) if Self::is_timeout(&e) => {
                    if filled == 0 {
                        return Ok(None); // No data at all — clean timeout
                    }
                    // Partial read — keep trying (data is in flight)
                    continue;
                }
                Err(e) => return Err(e.into()),
            }
        }
        Ok(Some(()))
    }

    /// Try to read one message from the port. Returns `None` immediately if no
    /// data is waiting (non-blocking fast-return path via `bytes_to_read`).
    pub fn try_read_event(&mut self) -> Result<Option<DelugeEvent>> {
        // Fast non-blocking check — avoid the 500ms port timeout when idle.
        match self.port.bytes_to_read() {
            Ok(n) if n < 3 => return Ok(None), // need at least 2-byte hdr + 1-byte type
            Err(_) => return Ok(None),
            _ => {}
        }

        // Read 2-byte length header
        let mut hdr = [0u8; 2];
        if self.read_exact_or_timeout(&mut hdr)?.is_none() {
            return Ok(None);
        }

        let msg_len = u16::from_le_bytes(hdr) as usize;
        if msg_len == 0 || msg_len > 1024 {
            bail!("Invalid message length: {msg_len}");
        }

        // Read type + data
        let mut payload = vec![0u8; msg_len];
        self.read_exact_or_timeout(&mut payload)?
            .context("Timeout reading message payload after header")?;

        let msg_type = payload[0];
        let data = &payload[1..];

        let Some(mt) = MsgFrom::from_byte(msg_type) else {
            bail!("Unknown message type: 0x{msg_type:02X}");
        };

        let event = match mt {
            MsgFrom::PadPressed if data.len() >= 2 => DelugeEvent::PadPressed {
                col: data[0],
                row: data[1],
            },
            MsgFrom::PadReleased if data.len() >= 2 => DelugeEvent::PadReleased {
                col: data[0],
                row: data[1],
            },
            MsgFrom::ButtonPressed if data.len() >= 1 => {
                DelugeEvent::ButtonPressed { id: data[0] }
            }
            MsgFrom::ButtonReleased if data.len() >= 1 => {
                DelugeEvent::ButtonReleased { id: data[0] }
            }
            MsgFrom::EncoderRotated if data.len() >= 2 => DelugeEvent::EncoderRotated {
                id: data[0],
                delta: data[1] as i8,
            },
            MsgFrom::EncoderPressed if data.len() >= 1 => {
                DelugeEvent::EncoderPressed { id: data[0] }
            }
            MsgFrom::EncoderReleased if data.len() >= 1 => {
                DelugeEvent::EncoderReleased { id: data[0] }
            }
            MsgFrom::Version if data.len() >= 3 => DelugeEvent::Version {
                major: data[0],
                minor: data[1],
                patch: data[2],
            },
            MsgFrom::Pong => DelugeEvent::Pong,
            MsgFrom::Ready => DelugeEvent::Ready,
            MsgFrom::Error => {
                let msg = String::from_utf8_lossy(data).to_string();
                DelugeEvent::Error(msg)
            }
            _ => bail!("Message 0x{msg_type:02X} with unexpected data length {}", data.len()),
        };

        Ok(Some(event))
    }

    // ── High-level commands ──────────────────────────────────────────

    /// Send a ping and wait for pong.
    pub fn ping(&mut self) -> Result<()> {
        self.send_raw(MsgTo::Ping as u8, &[])?;
        // Wait up to 1 second for pong (using multiple reads at the port's base timeout)
        let deadline = Instant::now() + Duration::from_secs(1);
        loop {
            if Instant::now() >= deadline {
                bail!("Ping timed out - no pong received");
            }
            match self.try_read_event()? {
                Some(DelugeEvent::Pong) => return Ok(()),
                Some(_) => continue, // skip other events
                None => continue,    // timeout on this read, retry until deadline
            }
        }
    }

    /// Request and return the firmware version.
    pub fn get_version(&mut self) -> Result<(u8, u8, u8)> {
        self.send_raw(MsgTo::GetVersion as u8, &[])?;
        let deadline = Instant::now() + Duration::from_secs(1);
        loop {
            if Instant::now() >= deadline {
                bail!("Get version timed out");
            }
            match self.try_read_event()? {
                Some(DelugeEvent::Version {
                    major,
                    minor,
                    patch,
                }) => return Ok((major, minor, patch)),
                Some(_) => continue,
                None => continue,
            }
        }
    }

    /// Set a single pad to an RGB color.
    #[allow(dead_code)]
    pub fn set_pad(&mut self, col: u8, row: u8, r: u8, g: u8, b: u8) -> Result<()> {
        self.send_raw(MsgTo::SetPadRgb as u8, &[col, row, r, g, b])
    }

    /// Set an orange indicator LED on or off (index 0–35).
    pub fn set_led(&mut self, index: u8, on: bool) -> Result<()> {
        self.send_raw(MsgTo::SetLed as u8, &[index, on as u8])
    }

    /// Set the 4-LED gold knob indicator bar.
    /// `which`: 0 = bottom knob (MOD_0), 1 = top knob (MOD_1).
    /// `levels`: brightness for each LED from bottom to top (0 = off, 255 = full).
    pub fn set_knob_indicator(&mut self, which: u8, levels: &[u8; 4]) -> Result<()> {
        self.send_raw(
            MsgTo::SetKnobIndicator as u8,
            &[which, levels[0], levels[1], levels[2], levels[3]],
        )
    }

    /// Set the tempo external-synced GPIO indicator LED.
    pub fn set_synced_led(&mut self, on: bool) -> Result<()> {
        self.send_raw(MsgTo::SetSyncedLed as u8, &[on as u8])
    }

    /// Turn off all button indicator LEDs, both gold knob indicator bars, and the synced LED.
    pub fn clear_all_leds(&mut self) -> Result<()> {
        self.send_raw(MsgTo::ClearAllLeds as u8, &[])
    }

    /// Clear all pads.
    pub fn clear_pads(&mut self) -> Result<()> {
        self.send_raw(MsgTo::ClearAllPads as u8, &[])
    }

    /// Send a full OLED framebuffer (768 bytes, 6 pages x 128 columns).
    /// Each byte represents 8 vertical pixels (LSB = top of page).
    pub fn update_display(&mut self, framebuffer: &[u8; OLED_FRAMEBUFFER_SIZE]) -> Result<()> {
        self.send_raw(MsgTo::UpdateDisplay as u8, framebuffer)
    }

    /// Clear the OLED display.
    pub fn clear_display(&mut self) -> Result<()> {
        self.send_raw(MsgTo::ClearDisplay as u8, &[])
    }

    /// Send all pad colours in one bulk message.
    /// `colors` must be col-major: for col in 0..PAD_COLS, row in 0..PAD_ROWS → [r, g, b].
    pub fn set_all_pads(&mut self, colors: &[[u8; 3]; PAD_COLS as usize * PAD_ROWS as usize]) -> Result<()> {
        let mut buf = [0u8; PAD_COLS as usize * PAD_ROWS as usize * 3];
        for (i, rgb) in colors.iter().enumerate() {
            buf[i * 3] = rgb[0];
            buf[i * 3 + 1] = rgb[1];
            buf[i * 3 + 2] = rgb[2];
        }
        self.send_raw(MsgTo::SetAllPads as u8, &buf)
    }
}
